iLiner: An Identity-by-Descent Analysis Tool
--------------------------------------------

iLiner is a pipeline for using iLASH.

To use iLiner, confirm the following are installed:

   - Python >= 3.5.0
   - iLASH

Then, on the command line, install with

.. code-block::

    .. pip install iliner
    git clone https://github.com/stephwen/iliner.git
    cd iliner
    python setup.py build
    python setup.py install

iLiner has three main modules.

   - Running iLASH
   - Calculating the IBD depth along the provided genome (all chromosomes provide)
   - Running statistical analysis on the data stratified by populations

Each module requires a separate set of commands.

The iLASH module can be run with two separate inputs.
  1. Phased Haplotype and Sample files
  2. Phased VCF

Haplotype and Sample Method Sample Code

.. code-block:: bash

  iliner ilash --sample /path/to/sample/file --haps /path/to/haplotype/file \
  --genetic_map /path/to/genetic/map/file --ilash /path/to/ilash --outfile_prefix \
  /output/path/with/prefix

VCF Method Sample Code

.. code-block::

  iliner ilash --vcf /path/to/sample/file --genetic_map /path/to/genetic/map/file \
  --ilash /path/to/ilash --outfile_prefix /output/path/with/prefix

Required Arguments:
  -ilash  This determines the module.
  --sample SAMPLE, -s SAMPLE  Location of the Sample File. This should be in PLINK format.

--haps HAPLOTYPE, -hp HAPLOTYPE
            Location of the Haplotype File. This should be in PLINK format and should be phase.

  --VCF VCF, -v VCF  Location of the VCF. This should phased and in standard genotype format.

--genetic_map GENTICMAP, -gm GENETICMAP
            Location of the genetic map. This should be for the same build as the genotype files.

  --ilash ILASH, -i ILASH  Location of the ilash executable.

  --outfile_prefix PREFIX, -op PREFIX  Pathway for the output file with prefix.

Optional Argument:
  --no_indel, -ni  This will remove indels.

The IBD depth module should use a file that contains iLASH results for all of the chromosomes of interest.

IBD Depth Sample Code

.. code-block::

  iliner ibd_depth --mapfile /path/to/map/file --ilash_output /path/to/ilash/output \
  --outfile_prefix /output/path/with/prefix

Required Flags:
  ibd_depth This determines the module.
  --mapfile, -mf  Location of a file containing all the chromosomes of interest
  --ilash_output, -io Location of a file containing the IBD results of interest
  --outfile_prefix, -op Pathway for the output files with prefix.

The statistics module should use the QCed version of the iLASH output generated by the IBD depth module.

Statistics Sample

.. code-block::

  iliner stats --ilash_output /path/to/qced/output --population_file /path/to/file/with/population/information --outfile_prefix /output/path/with/prefix

Required Flags:
  stats This determines the module.
  --ilash_output, -io Location of the QCed iLASH output from the IBD depth module.
  --population_file Location of the file with population information
  --outfile_prefix, -op Pathway for the output files with prefix.

Hannah Poisner <hannah.poisner@icahn.mssm.edu>
Stephane Wenric <stephane.wenric@mssm.edu>
